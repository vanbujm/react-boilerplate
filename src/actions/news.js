/* eslint-disable import/prefer-default-export */
import { REQUEST_REACT_NEWS } from '../constants/action-types/index';

export function requestReactNews() {
  return {
    type: REQUEST_REACT_NEWS,
  };
}
