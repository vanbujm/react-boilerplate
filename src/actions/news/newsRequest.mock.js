const mockResponse = (data, shouldResolve) => ({
  json: async () =>
    new Promise(resolve => {
      const response = shouldResolve ? { data } : { error: data };
      process.nextTick(() => resolve(response));
    }),
});

export default function newsRequest(mockNewsData, shouldResolve = true) {
  return new Promise(resolve => {
    if (shouldResolve) {
      process.nextTick(() =>
        resolve(mockResponse(mockNewsData, shouldResolve)),
      );
    } else {
      process.nextTick(() =>
        resolve(mockResponse('Everything broke', shouldResolve)),
      );
    }
  });
}
