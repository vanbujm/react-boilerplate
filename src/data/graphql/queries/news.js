import fetch from 'node-fetch';

// React.js News Feed (RSS)
const url =
  'https://api.rss2json.com/v1/api.json' +
  '?rss_url=https%3A%2F%2Freactjsnews.com%2Ffeed.xml';

let items = [];
let lastFetchTask;
let lastFetchTime = new Date(1970, 0, 1);

const news = () => {
  if (lastFetchTask) {
    return lastFetchTask;
  }

  const tenMinutes = 1000 * 60 * 10;

  if (new Date() - lastFetchTime > tenMinutes) {
    lastFetchTime = new Date();
    lastFetchTask = fetch(url)
      .then(response => response.json())
      .then(data => {
        if (data.status === 'ok') {
          items = data.items;
        }

        lastFetchTask = null;
        return items;
      })
      .catch(err => {
        lastFetchTask = null;
        throw err;
      });

    if (items.length) {
      return items;
    }

    return lastFetchTask;
  }

  return items;
};

export default news;
