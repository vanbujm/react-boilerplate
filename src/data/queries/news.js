/**
 * React Starter Kit (https://www.reactstarterkit.com/)
 *
 * Copyright Â© 2014-present Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import { GraphQLList as List } from 'graphql';
import fetch from 'node-fetch';
import NewsItemType from '../types/NewsItemType';

// React.js News Feed (RSS)
const url =
  'https://api.rss2json.com/v1/api.json' +
  '?rss_url=https%3A%2F%2Freactjsnews.com%2Ffeed.xml';

let items = [];

const news = {
  type: new List(NewsItemType),
  resolve(_, _1, _2, { cacheControl }) {
    const tenMinutes = 1000 * 60 * 10;
    cacheControl.setCacheHint({ maxAge: tenMinutes });

    return fetch(url)
      .then(response => response.json())
      .then(data => {
        if (data.status === 'ok') {
          items = data.items;
        }

        return items;
      })
      .catch(err => {
        throw err;
      });
  },
};

export default news;
