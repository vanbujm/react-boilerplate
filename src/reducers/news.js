import { SET_REACT_NEWS } from '../constants/action-types';

export default function news(state = {}, action) {
  switch (action.type) {
    case SET_REACT_NEWS:
      return {
        ...state,
        reactNews: action.news,
      };
    default:
      return state;
  }
}
