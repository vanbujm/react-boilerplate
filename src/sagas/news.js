import { takeEvery, call, put } from 'redux-saga/effects';
import { REQUEST_REACT_NEWS, SET_REACT_NEWS } from '../constants/index';
import createFetch from '../createFetch';
import { fetchNews } from '../common/fetchNews/fetchNews';

export function* retrieveReactNews() {
  const fetchFunction = createFetch(fetch, {
    baseUrl: window.App.apiUrl,
  });
  const news = yield call(fetchNews, fetchFunction);
  yield put({
    type: SET_REACT_NEWS,
    news,
  });
}

export function* watchRequestReactNews() {
  yield takeEvery(REQUEST_REACT_NEWS, retrieveReactNews);
}
