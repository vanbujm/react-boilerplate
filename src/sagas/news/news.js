import { takeEvery, call, put } from 'redux-saga/effects';
import {
  REQUEST_REACT_NEWS,
  SET_REACT_NEWS,
  RESPONSE_RECEIVED,
  REACT_NEWS_FEED_ERROR,
} from '../../constants/action-types/index';
import createFetch from '../../createFetch';
import { fetchNews } from '../../common/fetchNews/fetchNews';

export function* retrieveReactNews() {
  const fetchFunction = createFetch(fetch, {
    baseUrl: window.App.apiUrl,
  });

  const response = yield call(fetchNews, fetchFunction);

  yield put({ type: RESPONSE_RECEIVED });

  const getJson = async resp => resp.json();

  const { data } = yield call(getJson, response);

  if (!data || !data.news) yield put({ type: REACT_NEWS_FEED_ERROR });
  else {
    const { news } = data;
    yield put({
      type: SET_REACT_NEWS,
      news,
    });
  }
}

export function* watchRequestReactNews() {
  yield takeEvery(REQUEST_REACT_NEWS, retrieveReactNews);
}
