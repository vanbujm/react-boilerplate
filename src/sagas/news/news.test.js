/* eslint-env jest */
/* eslint-disable no-restricted-syntax */
import 'babel-polyfill';
import { runSaga } from 'redux-saga';
import { retrieveReactNews } from './news';

jest.mock('../../common/fetchNews/fetchNews');

const dispatchSpy = jest.fn();

const reduxStore = {
  dispatch: action => dispatchSpy(action),
  getState: () => ({ state: 'test' }),
};

describe('retrieveReactNews: ', () => {
  beforeEach(() => {
    window.App = {
      apiUrl: 'testURL',
    };
  });

  test('it dispatches the correct actions', async () => {
    await runSaga(reduxStore, retrieveReactNews).done;
    expect(dispatchSpy.mock.calls).toMatchSnapshot();
  });
});
