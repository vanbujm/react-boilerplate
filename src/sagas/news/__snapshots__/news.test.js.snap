// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retrieveReactNews:  it dispatches the correct actions 1`] = `
Array [
  Array [
    Object {
      "type": "RESPONSE_RECEIVED",
    },
  ],
  Array [
    Object {
      "news": Array [
        Object {
          "content": "<h1>Some mock content with html </h1>",
          "link": "https://reactjsnews.com/fake-news",
          "title": "Using Proxies with Redux Types",
        },
      ],
      "type": "SET_REACT_NEWS",
    },
  ],
]
`;
